### DISCUSSION TITLE
Object Merging Discussion

### DISCUSSION THREAD

**Discussion by rphsoftware** [Gilbert142/gomori#11] [General]
I feel like it would be beneficial to discuss with the community how object merging for each category of mergable object should be done.

I am aware of the following object types:
- RpgMaker Database entries (CommonEvents, Items, Armors, Skills etc)
- RpgMaker map data (data/Map<xyz>.json)
- Tiled Map data (maps/Map<xyz>.json)
- Translation files

Since each object type is slightly different, subtle differences to the merging strategies have to be made.

I would like to propose the following:
- RpgMaker database entries:
  - Conflict when 2 mods change the same entry in the database (if 2 mods modify, say, entry 50 in CommonEvents, there will be a conflict, even if the internal modifications are compatible)
- RpgMaker map data
  - Allow only 1 mod to modify the data section and each of the configuration options
  - Allow multiple mods to edit different events
  - Only allow one mod to add new events per map
- Tiled Map Data
  - Since the data arrays are fixed length, if the mods touch different areas, we can allow multiple to modify one map, however only if they impact different tiles. If 2 mods want to modify the same tile, conflict
- Translations
  - Allow multiple mods to edit different text sections of one file

**Comment by rphsoftware** [comment]
Primary dilema I have right now is if we should allow multiple mods to add events to the same map. Technically RPGMaker doesn't care about the ID of map-specific events (I think) so we could just, increment the ID independently and copy the object, however I'm not sure how the engine would handle 2 events on a single tile and such.

**Reply by rphsoftware** [reply]
Okay, this approach will not work, for now we will just conflict if 2 mods try to add new events to the same map.

**Comment by wesleylin** [comment]
We could possibly allow some overwriting with the mods and do it in dependency order (and alphabetical if on same dependency level). Aka the " If 2 mods want to modify the same tile, conflict" we could say if mod A depends on mod B and they both modify the same map first apply mod B, then apply mod A.
